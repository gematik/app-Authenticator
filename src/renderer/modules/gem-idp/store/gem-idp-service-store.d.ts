import { OAUTH2_ERROR_TYPE, TCallback } from '@/renderer/modules/gem-idp/type-definitions';

export type IOpenIdConfiguration = {
  authorization_endpoint: string;
  auth_pair_endpoint: string;
  sso_endpoint: string;
  uri_pair: string;
  token_endpoint: string;
  third_party_authorization_endpoint: string;
  uri_disc: string;
  issuer: string;
  jwks_uri: string;
  exp: number;
  iat: number;
  uri_puk_idp_enc: string;
  uri_puk_idp_sig: string;
  subject_types_supported: ['pairwise'];
  id_token_signing_alg_values_supported: ['BP256R1'];
  response_types_supported: ['code'];
  scopes_supported: ['openid', 'e-rezept', 'pairing'];
  response_modes_supported: ['query'];
  grant_types_supported: ['authorization_code'];
  acr_values_supported: ['gematik-ehealth-loa-high'];
  token_endpoint_auth_methods_supported: ['none'];
  code_challenge_methods_supported: ['S256'];
  kk_app_list_uri: string;
};

export type IIdpEncJwk = {
  use: string;
  kid: string;
  kty: string;
  crv: string;
  x: string;
  y: string;
};

export type IJweChallenge = {
  protected: string;
  iv: string;
  ciphertext: string;
  tag: string;
};

export type TGemIdpServiceStore = {
  sid?: string;
  jwsHbaSignature?: string;
  jwsSmcbSignature?: string;
  challengePath: string;
  idpHost: string;
  challenge: string;
  userConsent?: {
    requested_scopes: Record<string, string>;
    requested_claims: Record<string, string>;
  };
  jweChallenge: null | IJweChallenge;
  openIdConfiguration?: IOpenIdConfiguration;
  idpEncJwk?: IIdpEncJwk;
  errorShown: boolean;
  caughtAuthFlowError?: {
    errorCode: string;
    errorDescription?: string;
    oauthErrorType?: OAUTH2_ERROR_TYPE;
  };
  clientId: string;

  callback: TCallback | undefined;
  deeplink: string;
};
