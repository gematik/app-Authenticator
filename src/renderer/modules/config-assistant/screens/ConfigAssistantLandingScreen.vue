<!--
  - Copyright 2025, gematik GmbH
  -
  - Licensed under the EUPL, Version 1.2 or - as soon they will be approved by the
  - European Commission â€“ subsequent versions of the EUPL (the "Licence").
  - You may not use this work except in compliance with the Licence.
  -
  - You find a copy of the Licence in the "Licence" file or at
  - https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
  -
  - Unless required by applicable law or agreed to in writing,
  - software distributed under the Licence is distributed on an "AS IS" basis,
  - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either expressed or implied.
  - In case of changes by gematik find details in the "Readme" file.
  -
  - See the Licence for the specific language governing permissions and limitations under the Licence.
  -->

<template>
  <div class="home flex flex-col mt-[128px] ml-2 mr-2 justify-items-center gap-2 text-center">
    <img src="@/assets/logo_authenticator.png" class="object-contain pr-1 h-[156px]" alt="logo icon" />
    <h2 id="lblMainCenter" class="ml-20 mr-20">{{ $t('config_assistant.landing_text') }}</h2>
    <div class="mt-8">
      <button class="bt rounded-[4px] ml-[15px] w-60" type="button" @click="startAssistant(KONNEKTOR_VENDORS.Rise)">
        {{ $t('config_assistant.start_rise') }}
      </button>
      <button class="bt rounded-[4px] ml-[15px] w-60" type="button" @click="startAssistant(KONNEKTOR_VENDORS.Koco)">
        {{ $t('config_assistant.start_koco') }}
      </button>
      <button class="bt ml-[15px] w-60" type="button" @click="startAssistant(KONNEKTOR_VENDORS.Secunet)">
        {{ $t('config_assistant.start_secunet') }}
      </button>
    </div>
  </div>
</template>

<script setup lang="ts">
import { useRouter } from 'vue-router';
import { KONNEKTOR_VENDORS } from '@/@types/common-types';

const router = useRouter();
const startAssistant = (name: KONNEKTOR_VENDORS) => {
  router.push({ name: 'config-assistant', params: { konnektor: name, step: 1 } });
};
</script>
