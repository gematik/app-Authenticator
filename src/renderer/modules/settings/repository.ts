/*
 * Copyright 2023 gematik GmbH
 *
 * The Authenticator App is licensed under the European Union Public Licence (EUPL); every use of the Authenticator App
 * Sourcecode must be in compliance with the EUPL.
 *
 * You will find more details about the EUPL here: https://joinup.ec.europa.eu/collection/eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the EUPL is distributed on an "AS
 * IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the EUPL for the specific
 * language governing permissions and limitations under the License.ee the Licence for the specific language governing
 * permissions and limitations under the Licence.
 */

import dot from 'dot-object';
import Swal from 'sweetalert2';

// #!if MOCK_MODE === 'ENABLED'
import { DEVELOPER_OPTIONS, MOCK_CONNECTOR_CONFIG } from '@/renderer/modules/connector/connector-mock/mock-config';
// #!endif
import {
  APP_NAME,
  CONFIG_FILE_NAME,
  IPC_GET_PATH,
  IPC_READ_CREDENTIALS,
  IPC_SAVE_CREDENTIALS,
  PROCESS_ENVS,
  PRODUCT_NAME,
} from '@/constants';
import { logger } from '@/renderer/service/logger';
import {
  CHECK_UPDATES_AUTOMATICALLY_CONFIG,
  ENTRY_OPTIONS_CONFIG_GROUP,
  PROXY_SETTINGS_CONFIG,
  SENSITIVE_CONFIG_KEYS_LIST,
  TLS_AUTH_TYPE_CONFIG,
} from '@/config';
import { TlsAuthType } from '@/@types/common-types';
import { ERROR_CODES } from '@/error-codes';
import { UserfacingError } from '@/renderer/errors/errors';
import i18n from '@/renderer/i18n';

export interface TRepositoryData {
  [key: string]: number | string | boolean;
}

/**
 * Holds data in the cache to prevent reading the config file frequently
 */
let storedData: TRepositoryData = {};

export interface ISettingsRepository {
  save(data: TRepositoryData): void;

  load(forceReload?: boolean): TRepositoryData;

  clear(): void;

  exist(): boolean;

  /**
   * This only updates the cached config and does not change the config.json
   * This is relevant for making function tests on the fly
   */
  setWithoutSave(data: TRepositoryData): void;
}

export const INITIAL_STATE = {
  [CHECK_UPDATES_AUTOMATICALLY_CONFIG]: true,
  [PROXY_SETTINGS_CONFIG.AUTH_TYPE]: false,
  [PROXY_SETTINGS_CONFIG.USE_OS_SETTINGS]: true,
  [ENTRY_OPTIONS_CONFIG_GROUP.TLS_REJECT_UNAUTHORIZED]: false,
  [TLS_AUTH_TYPE_CONFIG]: TlsAuthType.ServerCertAuth,

  // #!if MOCK_MODE === 'ENABLED'
  [MOCK_CONNECTOR_CONFIG]: false,
  [DEVELOPER_OPTIONS.IDP_CERTIFICATE_CHECK]: true,
  // #!endif
};

export class FileStorageRepository implements ISettingsRepository {
  static get isNewInstallation(): boolean {
    return this._isNewInstallation;
  }
  static set isNewInstallation(value: boolean) {
    this._isNewInstallation = value;
  }

  static get usesCredentialManager(): boolean {
    return this._usesCredentialManager;
  }

  static set usesCredentialManager(value: boolean) {
    this._usesCredentialManager = value;
  }

  static get isJsonFileInvalid(): boolean {
    return this._isJsonFileInvalid;
  }

  static set isJsonFileInvalid(value: boolean) {
    this._isJsonFileInvalid = value;
  }

  private static encoding: BufferEncoding = 'utf-8';
  private static _path: string | null = null;
  private static _usesCredentialManager = false;
  private static _isNewInstallation = false;
  private static _isJsonFileInvalid = false;

  public static getConfigDir(): { path: string; localEnv: boolean } {
    const clientName = PROCESS_ENVS.CLIENTNAME;
    const computerName = PROCESS_ENVS.COMPUTERNAME;
    const authConfigPath = PROCESS_ENVS.AUTHCONFIGPATH;
    const viewClientMachineName = PROCESS_ENVS.VIEWCLIENT_MACHINE_NAME || PROCESS_ENVS.ViewClient_Machine_Name;

    logger.info('configPath:');
    logger.info('  -  AUTHCONFIGPATH:' + authConfigPath);
    logger.info('  -  CLIENTNAME:' + clientName);
    logger.info('  -  COMPUTERNAME:' + computerName);
    logger.info('  -  VIEWCLIENT_MACHINE_NAME:' + viewClientMachineName);

    // Abhängig von folgenden Umgebungsvariablen wird die Lokalität der zu verwendenden config.json in folgender Reihenfolge bestimmt:
    // 1. Wenn die Umgebungsvariable AUTHCONFIGPATH gesetzt ist und die Umgebungsvariable ViewClient_Machine_Name (VMWARE)
    //    → config.json im Folder: AUTHCONFIGPATH + //ViewClient_Machine_Name//config.json
    // 2. Wenn die Umgebungsvariable AUTHCONFIGPATH gesetzt ist und die Umgebungsvariable CLIENTNAME (CITRIX)
    //    → config.json im Folder: AUTHCONFIGPATH + //CLIENTNAME//config.json
    // 3. Wenn die Umgebungsvariable AUTHCONFIGPATH gesetzt ist und die Umgebungsvariable COMPUTERNAME
    //    → config.json im Folder: AUTHCONFIGPATH + //COMPUTERNAME//config.json
    // 4. Für alle anderen Fälle (auch im Fehlerfall das der AUTHCONFIGPATH nicht existiert!):
    //    → config.json im Installationsverzeichnis der exe (wie bisher)

    //wenn keine config.json in dem viewClientMachineName, clientName oder computerName Pfad existiert, dann wird die lokale config benutzt
    let configPath = '';
    if (authConfigPath) {
      if (viewClientMachineName) {
        logger.info(' - config über ViewClient_Machine_Name');
        configPath = window.api.pathJoin(authConfigPath.toString(), viewClientMachineName.toString());
      } else if (clientName) {
        logger.info(' - config über CLIENTNAME');
        configPath = window.api.pathJoin(authConfigPath.toString(), clientName.toString());
      } else if (computerName) {
        logger.info(' - config über COMPUTERNAME');
        configPath = window.api.pathJoin(authConfigPath.toString(), computerName.toString());
      }
      return { path: configPath, localEnv: false };
    }

    logger.info(' - config über Anwendungsverzeichnis');
    return { path: window.api.sendSync(IPC_GET_PATH, 'userData') as string, localEnv: true };
  }

  public static getPath(): string {
    if (FileStorageRepository._path) {
      return FileStorageRepository._path;
    }

    const config = FileStorageRepository.getConfigDir();
    let path = config.path;
    if (config.localEnv) {
      path = path.replace('Roaming', 'Local');

      //C:\Users\USERNAME\AppData\Local\authenticator -> C:\Users\USERNAME\AppData\Local\gematik Authenticator
      path = path.replace(window.api.pathSep() + APP_NAME, window.api.pathSep() + PRODUCT_NAME);

      if (!window.api.existsSync(path)) {
        try {
          window.api.mkdirSync(path);
        } catch (err) {
          logger.error('Config dir could not be created', err.message);
        }
      }
    }

    path += window.api.pathJoin('/', CONFIG_FILE_NAME);

    FileStorageRepository._path = path;

    logger.info(' - config.json path: ' + path);
    return path;
  }

  save(data: TRepositoryData): void {
    // update state
    storedData = dot.dot(data);

    // send data to preload & main
    window.api.setAppConfigInPreload(storedData);

    const { shouldSaveToCM, showWarningOnFail, saveAllToConfigFileOnFail } = FileStorageRepository.decideToSaveToCM();
    const nonSensitiveData = FileStorageRepository.filterSensitiveData(data);
    let dataToSaveConfigFile: TRepositoryData = saveAllToConfigFileOnFail ? data : nonSensitiveData;

    // save to credential manager
    if (shouldSaveToCM) {
      const savedToCMSuccessfully = FileStorageRepository.saveToCm(data);

      if (!savedToCMSuccessfully && showWarningOnFail) {
        throw new UserfacingError('Could not save to credential manager', '', ERROR_CODES.AUTHCL_0010);
      }

      if (savedToCMSuccessfully) {
        dataToSaveConfigFile = nonSensitiveData;
      }
    } else {
      dataToSaveConfigFile = data;
    }

    const unFlatted = dot.object({ ...dataToSaveConfigFile });
    window.api.writeFileSync(FileStorageRepository.getPath(), JSON.stringify(unFlatted, null, 2));

    logger.info('Saved config changes under: ', FileStorageRepository.getPath());

    // #!if MOCK_MODE === 'ENABLED'
    FileStorageRepository.printConfig();
    // #!endif
  }

  static saveToCm(data?: TRepositoryData): boolean {
    return window.api.sendSync(IPC_SAVE_CREDENTIALS, data) as boolean;
  }

  static readFromCm(): Partial<TRepositoryData> {
    return (window.api.sendSync(IPC_READ_CREDENTIALS) as Partial<TRepositoryData>) || {};
  }

  static decideToSaveToCM(): {
    shouldSaveToCM: boolean;
    showWarningOnFail: boolean;
    saveAllToConfigFileOnFail: boolean;
  } {
    let shouldSaveToCM;
    let showWarningOnFail;
    let saveAllToConfigFileOnFail;

    const isCentralConfiguration = PROCESS_ENVS.AUTHCONFIGPATH;

    // If it's a new installation
    if (FileStorageRepository._isNewInstallation) {
      shouldSaveToCM = true;
      showWarningOnFail = true;
      saveAllToConfigFileOnFail = false;
    } else {
      if (isCentralConfiguration) {
        if (FileStorageRepository._usesCredentialManager) {
          shouldSaveToCM = true;
          saveAllToConfigFileOnFail = false;
          showWarningOnFail = true;
        } else {
          shouldSaveToCM = false;
          saveAllToConfigFileOnFail = true;
          showWarningOnFail = true;
        }
      } else {
        saveAllToConfigFileOnFail = false;
        showWarningOnFail = true;
        shouldSaveToCM = true;
      }
    }

    return { shouldSaveToCM, showWarningOnFail, saveAllToConfigFileOnFail };
  }
  /**
   * This function filters sensitive data from the config data
   * @param data
   */
  static filterSensitiveData(data: TRepositoryData): TRepositoryData {
    const nonSensitiveData: TRepositoryData = {};

    for (const key in data) {
      if (!SENSITIVE_CONFIG_KEYS_LIST.includes(key)) {
        nonSensitiveData[key] = data[key];
      }
    }

    return nonSensitiveData;
  }

  /**
   * todo be sure that config file exists, and has a right format. If not, throw error
   */
  load(forceReload = false): TRepositoryData {
    if (!forceReload && Object.keys(storedData).length) {
      return storedData;
    }

    // forceReload happens when env vars get changed;
    // that means we need to recalculate the path
    if (forceReload) {
      FileStorageRepository._path = null;
    }

    const sensitiveDataFromCredentialsManager = FileStorageRepository.readFromCm();

    // set usesCredentialManager if sensitiveDataFromCredentialsManager is not empty
    FileStorageRepository._usesCredentialManager = !!Object.keys(sensitiveDataFromCredentialsManager).length;

    if (!window.api.existsSync(FileStorageRepository.getPath())) {
      FileStorageRepository._isNewInstallation = true;
      return { ...INITIAL_STATE };
    }

    const buffer = window.api.readFileSync(FileStorageRepository.getPath(), FileStorageRepository.encoding);
    const string = buffer.toString();

    let json: TRepositoryData;

    try {
      // convert string to json and convert string booleans to booleans
      json = JSON.parse(string, (_, value) => {
        if (value === 'true') return true;
        if (value === 'false') return false;
        return value;
      });
      FileStorageRepository._isJsonFileInvalid = false;
    } catch (e) {
      logger.error('Error while parsing config file', e);
      FileStorageRepository._isJsonFileInvalid = true;

      Swal.fire({
        icon: 'error',
        title: i18n.global.t(`errors.${ERROR_CODES.AUTHCL_0012}.title`),
        text: i18n.global.t(`errors.${ERROR_CODES.AUTHCL_0012}.text`, { path: FileStorageRepository.getPath() }),
      });

      return {};
    }

    const dottedData = {
      ...dot.dot(json),
      // get sensitive data from credential manager
      ...sensitiveDataFromCredentialsManager,
    };

    const data = {
      ...INITIAL_STATE,
      ...dottedData,
    };
    storedData = data;

    // send data to preload & main
    window.api.setAppConfigInPreload(storedData);
    // #!if MOCK_MODE === 'ENABLED'
    FileStorageRepository.printConfig();
    // #!endif
    return data;
  }

  clear(): void {
    storedData = {};

    if (window.api.existsSync(FileStorageRepository.getPath())) {
      window.api.unlinkSync(FileStorageRepository.getPath());
    }
  }

  exist(): boolean {
    return window.api.existsSync(FileStorageRepository.getPath());
  }

  /**
   * This only updates the cached config and does not change the config.json
   * @param data
   */
  setWithoutSave(data: TRepositoryData): void {
    storedData = data;
  }

  // #!if MOCK_MODE === 'ENABLED'
  private static printConfig() {
    logger.info('print settings:');
    for (const item in storedData) {
      if (!item.toLowerCase().includes('password')) {
        logger.info('  - ' + item + ':' + storedData[item]);
      }
    }
  }

  // #!endif
}
