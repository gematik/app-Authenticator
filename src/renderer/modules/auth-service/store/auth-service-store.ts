/*
 * Copyright 2023 gematik GmbH
 *
 * The Authenticator App is licensed under the European Union Public Licence (EUPL); every use of the Authenticator App
 * Sourcecode must be in compliance with the EUPL.
 *
 * You will find more details about the EUPL here: https://joinup.ec.europa.eu/collection/eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the EUPL is distributed on an "AS
 * IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the EUPL for the specific
 * language governing permissions and limitations under the License.ee the Licence for the specific language governing
 * permissions and limitations under the Licence.
 */

import { ActionContext, Module } from 'vuex';

import { TRootStore } from '@/renderer/store';
import { IAuthServiceStore } from './auth-service-store.d';
import { logger } from '@/renderer/service/logger';
import { KeycloakPluginError, UserfacingError } from '@/renderer/errors/errors';
import getIdpTlsCertificates from '@/renderer/utils/get-idp-tls-certificates';
import { TOidcProtocol2UrlSpec } from '@/@types/common-types';
import { validateAuthResponseData } from '@/renderer/service/idp-response-validators';
import { userAgent } from '@/renderer/utils/utils';

export const INITIAL_AUTH_STORE_STATE = {
  challengePath: '',
  challenge: '',
  authRequestPath: '',
};

const httpsReqConfig = () => ({
  https: {
    certificateAuthority: getIdpTlsCertificates(),
    rejectUnauthorized: true,
  },
  headers: {
    'Content-Type': 'application/json; charset=UTF-8',
    'User-Agent': userAgent,
  },
});

export const authServiceStore: Module<IAuthServiceStore, TRootStore> = {
  namespaced: true,
  state: (): IAuthServiceStore => INITIAL_AUTH_STORE_STATE,
  mutations: {
    setHbaJwsSignature(state: IAuthServiceStore, jwsSignature: string): void {
      state.jwsHbaSignature = jwsSignature;
    },
    setSmcbJwsSignature(state: IAuthServiceStore, jwsSignature: string): void {
      state.jwsSmcbSignature = jwsSignature;
    },
    setSid(state: IAuthServiceStore, sid: string): void {
      state.sid = sid;
    },
    setChallengePath(state: IAuthServiceStore, challengePath: string): void {
      state.challengePath = challengePath;
    },
    setChallenge(state: IAuthServiceStore, challenge: string): void {
      state.challenge = challenge;
    },
    resetStore(state: IAuthServiceStore): void {
      state.authRequestPath = undefined;
      state.jwsHbaSignature = undefined;
      state.jwsSmcbSignature = undefined;
      state.sid = undefined;
      state.challengePath = '';
      state.challenge = '';
    },
    setAuthRequestPath(state: IAuthServiceStore, authReqParameters: TOidcProtocol2UrlSpec): void {
      state.authRequestPath = authReqParameters.authz_path;
    },
  },
  actions: {
    async getChallengeData(context: ActionContext<IAuthServiceStore, TRootStore>): Promise<boolean> {
      const { state, commit } = context;
      logger.debug('state.authRequestPath ' + JSON.stringify(context));
      try {
        //  IDP host and Authorization Path
        const url = state.authRequestPath;
        const idpHost = new URL(<string>url);
        const { data } = await window.api.httpGet(<string>url, { ...httpsReqConfig(), followRedirect: true });

        logger.info('IdP: Challenge data received successfully.');

        // extract IDP Host without Authorization Path
        // Challenge Path:challenge_endpoint received from IDP Keycloak
        const challengePath: string = idpHost.protocol + '//' + idpHost.host + data.challenge_endpoint;

        /**
         * Validates Auth Response and throws error on invalid case
         */
        validateAuthResponseData(data);

        logger.info('IdP: Challenge data validation completed.');

        if (data.challenge_endpoint) {
          commit('setChallengePath', challengePath);
        }
        commit('setSid', data.sid);
        commit('setChallenge', data.challenge);

        return true;
      } catch (e: any) {
        // error comes from validateAuthResponseData, just forward it
        if (e instanceof UserfacingError) {
          throw e;
        }

        const response = e?.response;
        const errorMessage = response?.body?.error || e.message;
        logger.error(`Cannot get challenge from IdP. IdP responds with error: ${errorMessage}`, {
          error: e,
          status: response?.statusCode,
          data: response?.body,
          error_uri: response?.headers?.error_uri,
        });

        throw new KeycloakPluginError('Could not get challenge data for authentication', {
          error: response?.body?.error,
          url: response?.headers['error_uri'] || response?.request?.responseURL,
        });
      }
    },
    async getRedirectUriWithToken(
      context: ActionContext<IAuthServiceStore, TRootStore>,
    ): Promise<{ redirectUri: string; statusCode: number; error_uri?: string }> {
      const { state } = context;
      try {
        const payload = {
          hba: {
            signed_challenge: state.jwsHbaSignature,
          },
          smcb: {
            signed_challenge: state.jwsSmcbSignature,
          },
        };

        const response = await window.api.httpPost(state.challengePath, payload, httpsReqConfig());
        logger.debug(`IdP: Token received successfully. Http Status: ${response.status}`);

        return {
          redirectUri: <string>response?.headers?.['x-callback-location'],
          statusCode: response.status,
          error_uri: <string>response?.headers?.error_uri,
        };
      } catch (e: any) {
        const response = e?.response;
        logger.error('Could not get authorization code from IdP. Details: ', {
          error: e,
          status: response?.statusCode,
          data: response?.body,
          error_uri: response?.headers?.error_uri,
        });
        throw new KeycloakPluginError('Could not get authorization code from IdP. Details: ', {
          statusCode: response?.statusCode,
          error: response?.body?.error,
          url: response?.headers?.error_uri,
        });
      }
    },
  },
};
