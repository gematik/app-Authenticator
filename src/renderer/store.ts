/*
 * Copyright 2025, gematik GmbH
 *
 * Licensed under the EUPL, Version 1.2 or - as soon they will be approved by the
 * European Commission â€“ subsequent versions of the EUPL (the "Licence").
 * You may not use this work except in compliance with the Licence.
 *
 * You find a copy of the Licence in the "Licence" file or at
 * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either expressed or implied.
 * In case of changes by gematik find details in the "Readme" file.
 *
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 *
 * ******
 *
 * For additional notes and disclaimer from gematik and in case of changes by gematik find details in the "Readme" file.
 */

import { createStore } from 'vuex';
import { connectorStore } from '@/renderer/modules/connector/store/connector-store';
import { TConnectorStore } from '@/renderer/modules/connector/type-definitions';
import { TIdpServiceStore } from '@/renderer/modules/gem-idp/store/idp-service-store.d';
import { idpServiceStore } from '@/renderer/modules/gem-idp/store/idp-service-store';

export type TRootStore = {
  connectorStore: TConnectorStore;
  idpServiceStore: TIdpServiceStore;
  settingsIsSet: boolean;
  showLoadingSpinner: boolean;
};

export default createStore({
  state: (): TRootStore => {
    return {
      showLoadingSpinner: false,
    } as TRootStore;
  },
  mutations: {
    setShowLoadingSpinner(state, newState: boolean) {
      state.showLoadingSpinner = newState;
    },
  },
  actions: {},
  modules: {
    connectorStore,
    idpServiceStore,
  },
});
