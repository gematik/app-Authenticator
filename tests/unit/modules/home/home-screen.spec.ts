/*
 * Copyright 2024 gematik GmbH
 *
 * The Authenticator App is licensed under the European Union Public Licence (EUPL); every use of the Authenticator App
 * Sourcecode must be in compliance with the EUPL.
 *
 * You will find more details about the EUPL here: https://joinup.ec.europa.eu/collection/eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the EUPL is distributed on an "AS
 * IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the EUPL for the specific
 * language governing permissions and limitations under the License.ee the Licence for the specific language governing
 * permissions and limitations under the Licence.
 */

import { mount } from '@vue/test-utils';
import { HomeScreen } from '@/renderer/modules/home/screens';
import i18n from '@/renderer/i18n';
import store from '@/renderer/store';
import { FileStorageRepository } from '@/renderer/modules/settings/repository';

const fileStorageRepository = new FileStorageRepository();

jest.mock('@/renderer/modules/settings/useSettings.ts', () => ({
  useSettings: () => {
    return fileStorageRepository;
  },
}));

describe('home screen', () => {
  beforeEach(() => {
    fileStorageRepository.clear();
  });

  it('render with settings button', async function () {
    const wrapper = mount(HomeScreen, {
      global: {
        plugins: [store, i18n],
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('render with loading spinner', async function () {
    await store.commit('setShowLoadingSpinner', true);

    const wrapper = mount(HomeScreen, {
      global: {
        plugins: [store, i18n],
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
