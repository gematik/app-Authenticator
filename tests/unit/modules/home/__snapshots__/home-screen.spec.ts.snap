// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home screen render with loading spinner 1`] = `
<div
  class="home flex flex-col justify-center w-full mt-[128px]"
>
  <div
    class="flex row justify-center"
  >
    <img
      alt="logo icon"
      class="object-contain pr-1 h-[156px]"
    />
  </div>
  <h2
    class="text-center"
    id="lblMainCenter"
  >
    Sichere Anmeldung in der Telematikinfrastruktur
  </h2>
  <div
    class="flex justify-center"
  >
    <div
      class="loader"
    />
  </div>
  <div
    class="flex flex-col items-center mt-[64px]"
  >
    <p
      class="text-base text-baseText mb-[16px]"
    >
      Es wurden noch keine Einstellungen vorgenommen.
    </p>
    <button
      class="bt"
      type="submit"
    >
      <img
        alt="settings icon"
        class="inline pr-[8px]"
        src="[object Object]"
      />
      Einstellungen öffnen
    </button>
  </div>
  <!--v-if-->
</div>
`;

exports[`home screen render with settings button 1`] = `
<div
  class="home flex flex-col justify-center w-full mt-[128px]"
>
  <div
    class="flex row justify-center"
  >
    <img
      alt="logo icon"
      class="object-contain pr-1 h-[156px]"
    />
  </div>
  <h2
    class="text-center"
    id="lblMainCenter"
  >
    Sichere Anmeldung in der Telematikinfrastruktur
  </h2>
  <!--v-if-->
  <div
    class="flex flex-col items-center mt-[64px]"
  >
    <p
      class="text-base text-baseText mb-[16px]"
    >
      Es wurden noch keine Einstellungen vorgenommen.
    </p>
    <button
      class="bt"
      type="submit"
    >
      <img
        alt="settings icon"
        class="inline pr-[8px]"
        src="[object Object]"
      />
      Einstellungen öffnen
    </button>
  </div>
  <!--v-if-->
</div>
`;
